File: research/experiments/sweep/runs/opt-large-go-sonnet45/l_bfgs.go
package optimization

import "math"

// LBFGSOptions extends OptimizeOptions with L-BFGS specific parameters
type LBFGSOptions struct {
	OptimizeOptions
	Memory int
}

// DefaultLBFGSOptions returns default L-BFGS options
func DefaultLBFGSOptions(opts *LBFGSOptions) LBFGSOptions {
	options := LBFGSOptions{
		OptimizeOptions: DefaultOptions(nil),
		Memory:          10,
	}

	if opts != nil {
		if opts.GradTol != 0 {
			options.GradTol = opts.GradTol
		}
		if opts.StepTol != 0 {
			options.StepTol = opts.StepTol
		}
		if opts.FuncTol != 0 {
			options.FuncTol = opts.FuncTol
		}
		if opts.MaxIterations != 0 {
			options.MaxIterations = opts.MaxIterations
		}
		if opts.Memory != 0 {
			options.Memory = opts.Memory
		}
	}

	return options
}

// twoLoopRecursion computes H*g using the L-BFGS two-loop recursion
func twoLoopRecursion(g []float64, sHistory, yHistory [][]float64, rhoHistory []float64, gamma float64) []float64 {
	m := len(sHistory)

	q := Clone(g)
	alpha := make([]float64, m)

	// First loop (backward)
	for i := m - 1; i >= 0; i-- {
		alpha[i] = rhoHistory[i] * Dot(sHistory[i], q)
		q = AddScaled(q, yHistory[i], -alpha[i])
	}

	// Scale by gamma * I
	r := Scale(q, gamma)

	// Second loop (forward)
	for i := 0; i < m; i++ {
		beta := rhoHistory[i] * Dot(yHistory[i], r)
		r = AddScaled(r, sHistory[i], alpha[i]-beta)
	}

	return r
}

// LBFGS performs optimization using the limited-memory BFGS method
func LBFGS(
	f func([]float64) float64,
	x0 []float64,
	grad func([]float64) []float64,
	opts *LBFGSOptions,
) OptimizeResult {
	options := DefaultLBFGSOptions(opts)

	// If no gradient provided, use finite differences
	if grad == nil {
		grad = func(x []float64) []float64 {
			return ForwardDiffGradient(f, x)
		}
	}

	x := Clone(x0)
	fx := f(x)
	gx := grad(x)

	functionCalls := 1
	gradientCalls := 1

	// Check if already at minimum (use -1 to skip step and function checks)
	gradNorm := Norm(gx)
	reason := CheckConvergence(gradNorm, -1, -1, 0, options.OptimizeOptions)
	if reason != nil && IsConverged(reason) {
		return OptimizeResult{
			X:             x,
			Fun:           fx,
			Gradient:      gx,
			Iterations:    0,
			FunctionCalls: functionCalls,
			GradientCalls: gradientCalls,
			Converged:     true,
			Message:       ConvergenceMessage(reason),
		}
	}

	// History storage (circular buffer)
	var sHistory [][]float64
	var yHistory [][]float64
	var rhoHistory []float64
	gamma := 1.0

	for iter := 0; iter < options.MaxIterations; iter++ {
		// Compute search direction
		var d []float64
		if len(sHistory) == 0 {
			// First iteration: steepest descent
			d = Negate(gx)
		} else {
			// Use two-loop recursion
			Hg := twoLoopRecursion(gx, sHistory, yHistory, rhoHistory, gamma)
			d = Negate(Hg)
		}

		// Wolfe line search
		lsResult := WolfeLineSearch(f, grad, x, d, fx, gx, 1e-4, 0.9, 1e6, 25)
		functionCalls += lsResult.FunctionCalls
		gradientCalls += lsResult.GradientCalls

		if !lsResult.Success {
			return OptimizeResult{
				X:             x,
				Fun:           fx,
				Gradient:      gx,
				Iterations:    iter + 1,
				FunctionCalls: functionCalls,
				GradientCalls: gradientCalls,
				Converged:     false,
				Message:       ConvergenceMessage(&ConvergenceReason{Kind: "lineSearchFailed"}),
			}
		}

		// Update position
		xNew := AddScaled(x, d, lsResult.Alpha)
		fNew := lsResult.FNew
		gNew := lsResult.GNew

		// Compute step and gradient change
		s := Sub(xNew, x)
		y := Sub(gNew, gx)

		// Curvature condition
		ys := Dot(y, s)
		if ys > 1e-10 {
			rho := 1.0 / ys

			// Add to history
			sHistory = append(sHistory, s)
			yHistory = append(yHistory, y)
			rhoHistory = append(rhoHistory, rho)

			// Maintain circular buffer
			if len(sHistory) > options.Memory {
				sHistory = sHistory[1:]
				yHistory = yHistory[1:]
				rhoHistory = rhoHistory[1:]
			}

			// Update gamma
			yy := Dot(y, y)
			gamma = ys / yy
		}

		// Check convergence before updating
		gradNorm = Norm(gNew)
		stepNorm := Norm(s)
		funcChange := math.Abs(fNew - fx)

		// Update for next iteration
		x = xNew
		fx = fNew
		gx = gNew

		reason = CheckConvergence(gradNorm, stepNorm, funcChange, iter+1, options.OptimizeOptions)
		if reason != nil {
			return OptimizeResult{
				X:             x,
				Fun:           fx,
				Gradient:      gx,
				Iterations:    iter + 1,
				FunctionCalls: functionCalls,
				GradientCalls: gradientCalls,
				Converged:     IsConverged(reason),
				Message:       ConvergenceMessage(reason),
			}
		}
	}

	// Should not reach here
	return OptimizeResult{
		X:             x,
		Fun:           fx,
		Gradient:      gx,
		Iterations:    options.MaxIterations,
		FunctionCalls: functionCalls,
		GradientCalls: gradientCalls,
		Converged:     false,
		Message:       "Maximum iterations reached",
	}
}

File: research/experiments/sweep/runs/opt-full-rust-sonnet45/src/l_bfgs.rs
/// L-BFGS: limited-memory BFGS with two-loop recursion.

use crate::finite_diff::forward_diff_gradient;
use crate::line_search::wolfe_line_search;
use crate::result_types::{OptimizeOptions, OptimizeResult};
use crate::vec_ops::{dot, norm, sub};

fn two_loop_recursion(
    gx: &[f64],
    s_history: &[Vec<f64>],
    y_history: &[Vec<f64>],
    rho_history: &[f64],
    gamma: f64,
) -> Vec<f64> {
    let m = s_history.len();
    let n = gx.len();
    let mut q = gx.to_vec();
    let mut alphas = vec![0.0; m];

    // First loop: i = m-1 down to 0
    for i in (0..m).rev() {
        alphas[i] = rho_history[i] * dot(&s_history[i], &q);
        for j in 0..n {
            q[j] -= alphas[i] * y_history[i][j];
        }
    }

    // Scale: r = gamma * q
    let mut r: Vec<f64> = q.iter().map(|v| gamma * v).collect();

    // Second loop: i = 0 up to m-1
    for i in 0..m {
        let beta = rho_history[i] * dot(&y_history[i], &r);
        for j in 0..n {
            r[j] += (alphas[i] - beta) * s_history[i][j];
        }
    }

    r
}

pub fn lbfgs(
    f: &dyn Fn(&[f64]) -> f64,
    x0: &[f64],
    grad: Option<&dyn Fn(&[f64]) -> Vec<f64>>,
    options: Option<&OptimizeOptions>,
) -> OptimizeResult {
    lbfgs_with_memory(f, x0, grad, options, 10)
}

pub fn lbfgs_with_memory(
    f: &dyn Fn(&[f64]) -> f64,
    x0: &[f64],
    grad: Option<&dyn Fn(&[f64]) -> Vec<f64>>,
    options: Option<&OptimizeOptions>,
    memory: usize,
) -> OptimizeResult {
    let defaults = OptimizeOptions::default();
    let opts = options.unwrap_or(&defaults);

    let grad_fn = |x: &[f64]| -> Vec<f64> {
        match grad {
            Some(g) => g(x),
            None => forward_diff_gradient(f, x),
        }
    };

    let mut x = x0.to_vec();
    let mut fx = f(&x);
    let mut gx = grad_fn(&x);
    let mut function_calls = 1_usize;
    let mut gradient_calls = 1_usize;

    if norm(&gx) < opts.grad_tol {
        return OptimizeResult {
            x,
            fun: fx,
            gradient: Some(gx),
            iterations: 0,
            function_calls,
            gradient_calls,
            converged: true,
            message: "Converged: gradient norm below tolerance".to_string(),
        };
    }

    let mut s_history: Vec<Vec<f64>> = Vec::new();
    let mut y_history: Vec<Vec<f64>> = Vec::new();
    let mut rho_history: Vec<f64> = Vec::new();
    let mut gamma = 1.0_f64;

    for iteration in 1..=opts.max_iterations {
        // Direction
        let d: Vec<f64> = if s_history.is_empty() {
            gx.iter().map(|g| -g).collect()
        } else {
            let hg = two_loop_recursion(&gx, &s_history, &y_history, &rho_history, gamma);
            hg.iter().map(|v| -v).collect()
        };

        let ls = wolfe_line_search(f, &grad_fn, &x, &d, fx, &gx, None);
        function_calls += ls.function_calls;
        gradient_calls += ls.gradient_calls;

        if !ls.success {
            return OptimizeResult {
                x,
                fun: fx,
                gradient: Some(gx),
                iterations: iteration,
                function_calls,
                gradient_calls,
                converged: false,
                message: "Stopped: line search failed".to_string(),
            };
        }

        let x_new: Vec<f64> = x.iter().zip(d.iter()).map(|(xi, di)| xi + ls.alpha * di).collect();
        let g_new = ls.g_new.unwrap_or_else(|| grad_fn(&x_new));

        let s = sub(&x_new, &x);
        let y = sub(&g_new, &gx);
        let ys = dot(&s, &y);
        let step_norm = norm(&s);
        let func_change = (fx - ls.f_new).abs();

        x = x_new;
        fx = ls.f_new;
        gx = g_new;

        if ys > 1e-10 {
            if s_history.len() >= memory {
                s_history.remove(0);
                y_history.remove(0);
                rho_history.remove(0);
            }
            rho_history.push(1.0 / ys);
            s_history.push(s);
            y_history.push(y.clone());
            gamma = ys / dot(&y, &y);
        }

        let grad_norm = norm(&gx);
        if grad_norm < opts.grad_tol {
            return OptimizeResult {
                x,
                fun: fx,
                gradient: Some(gx),
                iterations: iteration,
                function_calls,
                gradient_calls,
                converged: true,
                message: format!("Converged: gradient norm {:.2e} below tolerance", grad_norm),
            };
        }
        if step_norm < opts.step_tol {
            return OptimizeResult {
                x,
                fun: fx,
                gradient: Some(gx),
                iterations: iteration,
                function_calls,
                gradient_calls,
                converged: true,
                message: "Converged: step size below tolerance".to_string(),
            };
        }
        if func_change < opts.func_tol {
            return OptimizeResult {
                x,
                fun: fx,
                gradient: Some(gx),
                iterations: iteration,
                function_calls,
                gradient_calls,
                converged: true,
                message: "Converged: function change below tolerance".to_string(),
            };
        }
    }

    OptimizeResult {
        x,
        fun: fx,
        gradient: Some(gx),
        iterations: opts.max_iterations,
        function_calls,
        gradient_calls,
        converged: false,
        message: format!("Stopped: reached maximum iterations ({})", opts.max_iterations),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_functions::*;

    #[test]
    fn test_sphere() {
        let tf = sphere();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!(r.fun < 1e-8);
    }

    #[test]
    fn test_booth() {
        let tf = booth();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!(r.fun < 1e-8);
    }

    #[test]
    fn test_rosenbrock() {
        let tf = rosenbrock();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!(r.fun < 1e-10);
    }

    #[test]
    fn test_beale() {
        let tf = beale();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!(r.fun < 1e-8);
    }

    #[test]
    fn test_himmelblau() {
        let tf = himmelblau();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!(r.fun < 1e-8);
    }

    #[test]
    fn test_goldstein_price() {
        let tf = goldstein_price();
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), None);
        assert!(r.converged);
        assert!((r.fun - 3.0).abs() < 1e-4);
    }

    #[test]
    fn test_sphere_fd() {
        let tf = sphere();
        let r = lbfgs(&tf.f, &tf.starting_point, None, None);
        assert!(r.converged);
        assert!(r.fun < 1e-6);
    }

    #[test]
    fn test_at_minimum() {
        let tf = sphere();
        let r = lbfgs(&tf.f, &[0.0, 0.0], Some(&tf.gradient), None);
        assert!(r.converged);
        assert_eq!(r.iterations, 0);
    }

    #[test]
    fn test_max_iter() {
        let tf = rosenbrock();
        let opts = OptimizeOptions { max_iterations: 2, ..Default::default() };
        let r = lbfgs(&tf.f, &tf.starting_point, Some(&tf.gradient), Some(&opts));
        assert!(!r.converged);
        assert!(r.message.contains("maximum iterations"));
    }

    #[test]
    fn test_custom_memory() {
        let tf = rosenbrock();
        let r = lbfgs_with_memory(&tf.f, &tf.starting_point, Some(&tf.gradient), None, 3);
        assert!(r.converged);
        assert!(r.fun < 1e-6);
    }
}
